[compiled autograd] Compiled autograd configs in TLS

xmfan opened the pull request:
Multithreaded doesn't work yet, this adds python side TLS only for the python side state

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #138061
* __->__ #137821
* #137953



cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @rec

xmfan committed:
[compiled autograd] Compiled autograd TLS

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] Compiled autograd TLS"




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] Compiled autograd TLS"




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] Compiled autograd TLS"




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] Compiled autograd TLS"




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] Compiled autograd TLS"




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] Compiled autograd TLS"




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] Compiled autograd configs in TLS"

Multithreaded doesn't work yet, this adds python side TLS for the configs passing




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] Compiled autograd configs in TLS"

Multithreaded doesn't work yet, this adds python side TLS for the configs passing




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

yf225 commented on test/distributed/_composable/fsdp/test_fully_shard_compile.py:
```diff
@@ -243,7 +243,7 @@ def _check_count(copy_count, resize_count):
             )
 
-        if not torch._dynamo.compiled_autograd.in_compiled_autograd_region:
+        if not torch._dynamo.compiled_autograd.local.get("in_compiled_autograd_region"):
```
Wonder should we encapsulate `in_compiled_autograd_region` into a function like `torch._dynamo.compiled_autograd.in_compiled_autograd_region()` to avoid needing to type `.local.` in user code? (similarly for `compiled_autograd_enabled`)?

(in my experiment enabling Traceable FSDP for Ads model I ended up needing check both flags, and simplifying them would be nice)

xmfan commented:
@pytorchbot merge

wdvr commented:
@pytorchmergebot --help

wdvr commented:
@pytorchbot revert -m 'Reverting this for now, it is failing test_public_bindings in trunk' -c weird

test_public_bindings.py::TestPublicBindings::test_correct_module_names [GH job link](https://github.com/pytorch/pytorch/actions/runs/11364250964/job/31612120389) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/15722debfb95863d666ca88ff6ac98a57f89c4eb)

xmfan committed:
Update on "[compiled autograd] Compiled autograd configs in TLS"

Multithreaded doesn't work yet, this adds python side TLS only for the python side state




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] Compiled autograd configs in TLS"

Multithreaded doesn't work yet, this adds python side TLS only for the python side state




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] Compiled autograd configs in TLS"

Multithreaded doesn't work yet, this adds python side TLS only for the python side state




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] Compiled autograd configs in TLS"

Multithreaded doesn't work yet, this adds python side TLS only for the python side state




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]
